game-core/src/main/java/games/strategy/triplea/ui/TripleAFrame.java

```diff
@@ -596,38 +599,62 @@ public void mouseEntered(final Territory territory) {
     }
 
     void refresh() {
+      territoryInfo.removeAll();
+
       final StringBuilder buf = new StringBuilder();
       buf.append(territoryLastEntered == null ? "none" : territoryLastEntered.getName());
       if (territoryLastEntered != null) {
         final TerritoryAttachment ta = TerritoryAttachment.get(territoryLastEntered);
         if (ta != null) {
-          final List<TerritoryEffect> territoryEffect = ta.getTerritoryEffect();
-          if (!territoryEffect.isEmpty()) {
-            buf.append(territoryEffect.stream()
-                .map(TerritoryEffect::getName)
-                .collect(Collectors.joining(", ", " (", ")")));
+          final List<TerritoryEffect> territoryEffects = ta.getTerritoryEffect();
+          int count = 0;
+          final StringBuilder territoryEffectText = new StringBuilder();
+          for (final TerritoryEffect territoryEffect : territoryEffects) {
+            try {
+              final JLabel territoryEffectLabel = new JLabel();
+              territoryEffectLabel.setIcon(uiContext.getTerritoryEffectImageFactory().getIcon(territoryEffect, false));
+              territoryEffectLabel.setBorder(BorderFactory.createEmptyBorder(0, 0, 0, 10));
+              territoryInfo.add(territoryEffectLabel,
+                  new GridBagConstraints(count++, 0, 1, 1, 0, 0, GridBagConstraints.WEST, GridBagConstraints.NONE,
```

Comment:
Very much recommend to use grid bag helper, it's usage is mostly transparent and removes all of the complexities, just need to specify how many columns you want.

Reply:
Probably though I just adjusted the existing code which was using the grid bag classes directly. I think there is a lot of UI code that could be simplified with some of the helper swing classes that have been created. Though given @RoiEXLab push to most to JavaFX, I also don't know how much time investment in refactoring swing code I'd want to make.

Reply:
> Though given @RoiEXLab push to most to JavaFX, I also don't know how much time investment in refactoring swing code I'd want to make.

I would hate for that to be the reason to just keep going with dirty swing code. JavaFX has been in the works for a while, we're not past the initial lobby screens, the game engine client proper and purchase window are some time down the road. 

>  I think there is a lot of UI code that could be simplified with some of the helper swing classes that have been created

Indeed, part of the blocker for JavaFX is that the UI code also has a ton of model code in it. Once we simplify with builder, that tends to encourage a split, the conversion to JavaFX is then simpler. So it's a bit funny that converting to simpler builder is low value as it's then trivial to move to JavaFX, ironic a bit as it's required to do the simplification first to get JavaFX wired in easily.